### ISSUE TITLE
Toxcore should make more liberal use of assertions

### ISSUE THREAD

**Issue by iphydf** [TokTok/toxcore#8]
@JFreegman mentioned an [article](http://blog.regehr.org/archives/1091) about assertions that contains some useful thoughts on how to write good assertions and what to pay attention to.

As discussed in #5, artificially restricting the interface of a function to include sanity checks about layers on top of it is not ideal, since it breaks module abstractions. Also, swallowing severe code errors tends to hide bugs that could go unnoticed for some time.

Therefore, I would suggest we make a gradual move towards having our APIs accept their whole function domain and encoding bug-checking checks with knowledge of higher layers as assertions.

For a specific first case and its discussion, see #5.

**Comment by JFreegman** [comment]
Assuming we will be using asserts in production code, we need to take into consideration the impact they will have on both performance and UX. That generally means using them sparingly enough that non-serious/handleable bugs won't trigger them, and strategically enough that we won't see a noticeable decline in performance, while still maintaining justification of their usage. Basically, I don't want to see people throwing asserts around carelessly, in place of proper error handling.

So based on these requirements (which are of course up for debate), what are the criteria for when to use an assert?

**Comment by GrayHatter** [comment]
I'm assuming that production code will get `-DNDEBUG` so there wouldn't be any loss of performance.

Is that not the intent?

**Comment by iphydf** [comment]
Perhaps we can have two kinds of assert, one for more expensive checks that are not part of the production build, and another for cheap checks that are on by default in production builds.

Here are some of the things we should have checks for (assert or otherwise):

1. Violation of function contract. Passing NULL where it's not expected would constitute a contract violation. In file-scope functions, it's a judgement call. In global-scope functions, it makes sense to handle it gracefully, i.e. returning error instead of crashing. This would not be a good candidate for assert (and there I disagree somewhat with the article, which would use assert for that).
2. Corrupt data structures. These can be more expensive, as they may perform "fsck" style checks on data structures. It's unlikely that we currently have these kinds of checks, and they would be a candidate for an [xassertdb](https://github.com/panyam/elkhound/blob/master/smbase/xassert.h#L33)-like macro.
3. Caller behaviour. This is encoding problem domain knowledge and this is what #5 is changing. The function shouldn't know that all callers must exhibit a certain behaviour. If we want to check that the callers are doing the right thing, an assert makes sense. For example, the function #5 is changing has a check for length > 0, but not for length < 1500 (or some appropriate value < 1500 for which there is a constant), which would make as much sense.

**Comment by iphydf** [comment]
An alternative would be to leave the checks as they are, but mark spec-violating ones somehow, so they can be conditionally compiled to true, and in production builds they would be evaluated. This way we could have one "safer" and one "strictly compliant" build. The latter would be used for testing.

**Comment by neduard** [comment]
I am in strong favour of having as many asserts as possible, but used as mentioned in the article by John Regehr. In the section "Where Do Assertions Come From?" it is said they exist specifically to denote:
* pre conditions
* post conditions
* invariants
* *spec*
The benefit is that if there are many assertions, static analysers could be used at a later stage and start discharging them. Alternatively, the asserts could be used as a basis to start adding contracts in the source code. These can be enforced by static checkers such as frama-c / tis-interpreter. There will be no need to enable them in production as we will have a guarantee that those asserts will never be tripped and thus no performance penalty (they might even speed up execution as the `if` statements become shorter).

**Comment by JFreegman** [comment]
Both @iphydf and @neduard have made some good suggestions about usage. Let's finalize our decision to use them, and continue discussing methods. Anyone oppose? @irungentoo @nurupo

**Comment by irungentoo** [comment]
I agree with adding asserts for debugging what I don't agree with is replacing checks that are already there with them like what #5 is doing.

**Comment by iphydf** [comment]
Ok, so here is my current suggestion (implemented in the most recent version of #5 at the time of writing):
- `assert` for detectable but unrecoverable errors that would cause undefined behaviour if not checked (likely used sparsely).
- `assertdb` for detectable corruptions in data structures.
- `return_val_if_fail` and `return_if_fail` (like the glib macros) for checks that are not necessary at the given abstraction level, but in the broader context of toxcore make sense to have.
- `if + return -1` for spec-compliant input validation that any correct toxcore implementation must perform.

The first three are checks that could be elided if all (or most) of toxcore were formally verified using a frama-c style (sound) checker. The last kind is all the checks that somehow depend on user input, i.e. network traffic or public API calls. These checks will never be elided and the testsuite will actually validate that the implementation performs them.

**Comment by nbraud** [comment]
@iphydf This seems sane.

One quick question, though: what is the use-case for `DISABLE_CHECKS`?

**Comment by iphydf** [comment]
@nbraud the use case is disabling checks that make the code not conform to the conditions we would like it to. The encryption function can sensibly encrypt zero-length strings. In the context of toxcore, there should be no call that actually does it, which is why I initially suggested making it an assert-like failure (abort program, *serious* bug detected). Still, making the function *specified* to be partial when it can do sane things with a value from its domain is wrong at this level.

I would turn the null checks into assertions, as well, though of a different kind. No code has any business calling the encryption function with null values, and I would consider null not to be part of the specified interface. It can not do sane things with a null key, so it should not pretend it can. If it does, there is a bug elsewhere we want to know about. @irungentoo disagrees with this, and opts for the graceful degradation route. Right now, it's very hard (almost impossible, without doing abstract interpretation of some kind, i.e. a full audit of all paths leading up to the call) to know whether anything triggers the error condition.

Perhaps we can have a `DEFENSIVE` (name up for debate) macro that is used like:
```c
if (DEFENSIVE(length == 0)) {
  return -1; // or do whatever else to recover from the error and continue working.
}
```

The macro would be something like:
```c
#ifdef DISABLE_DEFENSIVE_CHECKS
#define DEFENSIVE(cond) false
#else
#define DEFENSIVE(cond) defensive_check(__FILE__, __LINE__, __func__, #cond, cond)
#endif

bool defensive_check(char const *file, unsigned int line, char const *func, char const *msg, bool cond);
```

Inside `defensive_check`, we can do different things depending on configuration. We could log failures and optionally abort the program.

Bottom line: I want checks for everything, but I want to make sure we mark the different types of checks clearly. Right now, toxcore mixes domain checks (`x != null`), defensive checks (`length != 0`), and dynamic value checks (decryption failed) and they all return -1. There is no way to know what happened and where.

**Comment by nbraud** [comment]
@iphydf OK, that makes sense.

**Comment by iphydf** [comment]
Where are we with this? Which route should we choose with these checks?

**Comment by iphydf** [comment]
This discussion has reached the end with no consensus. We'll revisit the issue when it becomes one.
